Lib/ntpath.py

```diff
@@ -640,23 +638,65 @@ def commonpath(paths):
         genericpath._check_arg_types('commonpath', *paths)
         raise
 
+MAX_PATH = 260
 
-# determine if two files are in fact the same file
 try:
-    # GetFinalPathNameByHandle is available starting with Windows 6.0.
-    # Windows XP and non-Windows OS'es will mock _getfinalpathname.
-    if sys.getwindowsversion()[:2] >= (6, 0):
-        from nt import _getfinalpathname
-    else:
-        raise ImportError
-except (AttributeError, ImportError):
-    # On Windows XP and earlier, two files are the same if their absolute
-    # pathnames are the same.
-    # Non-Windows operating systems fake this method with an XP
-    # approximation.
-    def _getfinalpathname(f):
-        return normcase(abspath(f))
+    from nt import _getfinalpathname
+
+    # determine if two files are in fact the same file
+    def realpath(filename):
+        filename = os.fspath(filename)
+        is_str = isinstance(filename, str)
+        if not is_str:
+            filename = os.fsdecode(filename)
+        extended_path_prefix = '\\\\?\\'
+        is_extended_path = filename.startswith(extended_path_prefix)
+        unresolved = filename if is_extended_path else abspath(filename)
+        resolved_parts = []
+        while True:
+            try:
+                resolved_parts.append(_getfinalpathname(unresolved))
+                break
+            except OSError:
+                unresolved, tail = split(unresolved)
+                if not tail:
+                    resolved_parts.append(unresolved)
+                    break
+                resolved_parts.append(tail)
+        resolved = join(*reversed(resolved_parts))
+        # try to convert extended path to normal if
+        # initial path did not use \\?\ prefix and result uses it
+        if not is_extended_path and resolved.startswith(extended_path_prefix):
+            resolved = _extended_to_normal(resolved)
+        return resolved if is_str else os.fsencode(resolved)
+
+    def _extended_to_normal(path):
+        letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
+        if path[5:7] == ":\\" and path[4].upper() in letters:
+            # extended path with \\?\ prefix
+            # 4 is len('\\?\')
+            normal_path = normpath(path[4:])
+        elif path[:8].upper() == '\\\\?\\UNC\\':
+            # UNC path with \\?\ prefix - drop prefix
+            # 7 is len('\\?\UNC')
+            normal_path = normpath('\\' + path[7:])
+        else:
+            # not a UNC or drive-letter path
+            # return path as-is
+            return path
 
+        if (len(normal_path) < MAX_PATH and
```

Comment:
I'd rather not check the length here and just handle the `OSError` that `_getfullpathname` raises if the path is too long (which may be 32k instead of 260 on more recent systems)

Reply:
That would be convenient, but `GetFullPathNameW` only has the NT limit for a null-terminated string, `UNICODE_STRING_MAX_CHARS - 1` (32766).  Someone made a couple of mistakes in the [docs](https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-getfullpathnamew).  Windows 10 long-path support has no bearing on `GetFullPathNameW`, and we've never had to use the "\\\\?\\" prefix to support up to 32,766 characters with this function. The only change I recall is that prior to Windows 7 the behavior wasn't well-defined if we passed it a path that exceeded 32,766 characters. I think it would type cast the length in bytes to `USHORT`. Since Windows 7, it calls the internal function `RtlGetFullPathName_UEx`, which returns an `NTSTATUS` code. If the input path length exceeds 32,766, `RtlGetFullPathName_UEx` fails with `STATUS_NAME_TOO_LONG` (i.e. `ERROR_FILENAME_EXCED_RANGE`). If the result would exceed 32,766 characters, it fails with `STATUS_OBJECT_NAME_INVALID` (i.e. `ERROR_INVALID_NAME`).

Reply:
> That would be convenient, but …

I don't understand why you say "but"? It sounds like (and I confirmed from source) all our supported platforms can handle normal and excessively long paths here just fine, so there's no need for the check, correct?

Reply:
We're probably talking past each other here. IMO, we shouldn't remove the "\\\\?\\" prefix for long paths prior to Windows 10, and in Windows 10 we shouldn't remove it if long-path support isn't enabled by the "LongPathsEnabled" value in "HKLM\SYSTEM\CurrentControlSet\Control\FileSystem". Maybe you mean that we can remove the prefix because none of our code will crash on long paths. My basis for suggesting this check was that `realpath` should return a usable result if possible, not a result that requires converting back to an extended path (including the special case for UNC paths) if long paths aren't enabled. 

We can skip the `MAX_PATH` check if we have a way to check whether long paths are enabled. The only way to directly check this is to call the undocumented function `RtlAreLongPathsEnabled`, if it's defined in ntdll.dll. We could also check the Windows version number and "LongPathsEnabled" registry value at startup, but it's slightly race-prone since the value may have changed after the system read and cached it at startup. I suppose we could check it indirectly by trying to open a random 255-character name in the temp folder at startup, if that's at least `MAX_PATH` characters. If it fails with `ERROR_FILE_NOT_FOUND` then set a flag that long paths are enabled. If long paths aren't enabled I expect the error to be `ERROR_PATH_NOT_FOUND`. 

I was also concerned about sharing paths via IPC and configuration settings with programs that don't support long paths. But now that I think about it more, I don't think extended paths help this case much in general since many programs don't support them. In this case we can use some type of junction, which could be a drive-letter junction via `win32file.DefineDosDevice` (or equivalently subst.exe) or a filesystem junction via `_winapi.CreateJunction` (or equivalently CMD's `mklink /j`). These extend the limit to about 4k characters. It's not the full 32k, but I have no idea who would really need paths that long.

Reply:
Ah, I see what you're thinking now.

I'd rather move the path length condition out of this function then, so that:
* `_extended_to_normal` _always_ returns a normal path, regardless of length
* `realpath` returns a normal path if < 260 chars or an extended path if > 260 chars

The first point is the main part that was bothering me.

Reply:
And `_extended_to_normal` should also have the prefix check, being a no-op if the path doesn't have it.